Add support for inferring tr() descriptors

sipa opened the pull request:
Includes:
* First commit from #21365, adding TaprootSpendData in SigningProvider
* A refactor to expose ComputeTapleafHash and ComputeTaprootMerkleRoot from script/interpreter
* A tiny change to make `getaddressinfo` report tr() descriptors as solvable (so that inferred descriptors are shown), despite not having signing code for them.
* Logic to infer the script tree back from TaprootSpendData, and then use that to infer descriptors.

sipa committed:
Taproot descriptor inference

sipa committed:
consensus refactor: extract ComputeTapleafHash, ComputeTaprootMerkleRoot

sipa committed:
Report address as solvable based on inferred descriptor

achow101 reviewed (commented):
ACK 77466ae5c58b4a46e545d3d24521fa635f7b8fbe

sipa commented on src/script/standard.cpp:
```diff
@@ -516,3 +516,120 @@ TaprootSpendData TaprootBuilder::GetSpendData() const
+                   (CHashWriter{HASHER_TAPBRANCH} << node.sub[1]->hash << node.sub[1]->hash).GetSHA256() == node.hash) {
+            // Special case to deal with duplicate subtrees: after exploring the left subtree (which holds actual data),
+            // iterate through it again, but mark the right (unexplored, but implicitly identical) subtree as finished,
+            // so that we don't end up in an infinite loop.
```
Restructured things a bit by adding explicit "explored" and "inner" flags to the nodes. This avoids the need for testing the specifics of the unexplored child.

Also expanded the comment to explain this better.

achow101 commented:
ACK f9d6f9a52c61833bbc5c611ff6a3e59066e1c92f

laanwj commented:
Reviewed that the consensus commit (e3739fb3b042c3805b5b6357f6f79df787d11144) is a pure refactor.

achow101 commented:
re-ACK 13c6ab699eb6906c6554b4319a377c33fa0d157f

laanwj commented on src/script/descriptor.cpp:
```diff
@@ -1217,44 +1219,68 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
 
+std::unique_ptr<PubkeyProvider> InferXOnlyPubkey(const XOnlyPubKey& xkey, ParseScriptContext ctx, const SigningProvider& provider)
+{
+    unsigned char full_key[33] = {0x02};
```
I would have a slight preference to use constants here instead of hardcoding `0x02` and `33`.

sipa replied:
I agree with this, but prefer keeping it for a follow-up. There are several places where this is already done, and I'd like to introduce constants for it, as well as using them everywhere in the same PR - doing so here would be a distraction.

laanwj commented on src/script/descriptor.cpp:
```diff
@@ -1217,44 +1219,68 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
 }
 
-std::unique_ptr<PubkeyProvider> InferPubkey(const CPubKey& pubkey, ParseScriptContext, const SigningProvider& provider)
+std::unique_ptr<PubkeyProvider> InferPubkey(const CPubKey& pubkey, ParseScriptContext, const SigningProvider& provider, bool xonly = false)
```
It looks like the optional `xonly` argument is never passed?

sipa replied:
Fixed.

laanwj commented on src/script/interpreter.cpp:
```diff
@@ -1868,8 +1866,19 @@ static bool VerifyTaprootCommitment(const std::vector<unsigned char>& control, c
+static bool VerifyTaprootCommitment(const std::vector<unsigned char>& control, const std::vector<unsigned char>& program, const uint256& tapleaf_hash)
+{
+    //! The internal pubkey (x-only, so no Y coordinate parity).
+    const XOnlyPubKey p{uint256(std::vector<unsigned char>(control.begin() + 1, control.begin() + TAPROOT_CONTROL_BASE_SIZE))};
```
would be nice to add an assertion here with regard to `control.size()` before indexing into it

sipa replied:
Done.

Sjors replied:
I think you lost the asserts in your rebase:
```
assert(control.size() >= TAPROOT_CONTROL_BASE_SIZE);
assert(program.size() >= uint256::size());
```

jonatack commented:
Reviewing tonight/tomorrow.

sipa commented:
Addressed some comments, and also made this observable change:

```diff
             if (permit_uncompressed || key.IsCompressed()) {
                 CPubKey pubkey = key.GetPubKey();
                 out.keys.emplace(pubkey.GetID(), key);
-                return std::make_unique<ConstPubkeyProvider>(key_exp_index, pubkey);
+                return std::make_unique<ConstPubkeyProvider>(key_exp_index, pubkey, ctx == ParseScriptContext::P2TR);
             } else {
```

Which will make descriptors with a provided private key inside `tr()` be interpreted as the equivalent xonly pubkey rather than the compressed or uncompressed one.

achow101 commented:
re-ACK 834935511c2c799c8cb2312d617dbed6336000c4

Sjors commented:
It's too hot here to review code, but I can confirm that this correctly infers my `tr(xpub...)` and `tr(xpub1, pk(xpub2))` descriptors (and surviving a roundtrip through `deriveaddresses`). Might be a good sanity check to try with a rebased #21329 too.

laanwj commented:
> It's too hot here to review code

Same, I propose we postpone the feature freeze a bit. Not able to concentrate at the moment.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -1282,6 +1306,40 @@ std::unique_ptr<DescriptorImpl> InferScript(const CScript& script, ParseScriptCo
+                        subdesc = InferScript(script, ParseScriptContext::P2TR, provider);
+                    }
+                    if (!subdesc) {
+                        ok = false;
```
834935511c2c799c8cb2312d617dbed6336000c4: rather than failing, could we return `raw()` (at least for `leaf_ver == TAPROOT_LEAF_TAPSCRIPT`)? That can wait for a followup though, since we can't have `tr()` descriptors with raw leaves in the wallet anyway atm, and `raw()` is still limited to the top level.

sipa replied:
I think that's exactly what this code does.

Sjors reviewed (approved):
tACK 8349355

However I glanced over `InferTaprootTree`. I think it would help to include tests for the various problems that function is checking for.

sipa commented on src/script/standard.cpp:
```diff
@@ -464,8 +490,33 @@ TaprootBuilder& TaprootBuilder::Finalize(const XOnlyPubKey& internal_key)
 
 WitnessV1Taproot TaprootBuilder::GetOutput() { return WitnessV1Taproot{m_output_key}; }
+
+TaprootSpendData TaprootBuilder::GetSpendData() const
```
Done in https://github.com/bitcoin/bitcoin/pull/22275.

sipa commented on src/script/standard.h:
```diff
@@ -209,15 +210,38 @@ CScript GetScriptForRawPubKey(const CPubKey& pubkey);
+    XOnlyPubKey internal_key;
+    /** The Merkle root of the script tree (0 if no scripts). */
+    uint256 merkle_root;
+    /** Map from (script, leaf_version) to (sets of) control blocks. */
```
Doing so in https://github.com/bitcoin/bitcoin/pull/22275.

sipa commented:
Rebased, will work on adding some tests for failure cases of `InferTaprootTree`.

Sjors reviewed (commented):
I think something went wrong in your rebase.

For example in the original https://github.com/bitcoin/bitcoin/commit/834935511c2c799c8cb2312d617dbed6336000c4 you changed the argument `bool xonly = false` to `bool xonly` in the `ConstPubkeyProvider` provider.

sipa commented:
@Sjors Oops, thanks for catching that. I redid the rebase; does it look better now?

achow101 commented:
re-ACK d637a9b397816e34652d0c4d383308e39770737a

Rebase looks good.

Sjors commented:
re-utACK d637a9b

Much better

meshcollider commented on src/script/standard.cpp:
```diff
@@ -520,3 +520,138 @@ TaprootSpendData TaprootBuilder::GetSpendData() const
+    std::vector<std::tuple<int, CScript, int>> ret;
+    if (spenddata.merkle_root.IsNull()) return ret;
+
+    /** Data structure to represent the nodes of the tree we're going to be build. */
```
nit: `to build`

jonatack replied:
touched up the various nits in #22323, feel free to add any others there

meshcollider commented on src/script/standard.cpp:
```diff
@@ -520,3 +520,138 @@ TaprootSpendData TaprootBuilder::GetSpendData() const
+
+    /** Data structure to represent the nodes of the tree we're going to be build. */
+    struct TreeNode {
+        /** Hash of this none, if known; 0 otherwise. */
```
nit: `node`

meshcollider commented on src/script/standard.cpp:
```diff
@@ -520,3 +520,138 @@ TaprootSpendData TaprootBuilder::GetSpendData() const
+        bool done = false;
+    };
+
+    // Build tree from the provides branches.
```
nit: `provided`

meshcollider reviewed (commented):
Code review ACK d637a9b397816e34652d0c4d383308e39770737a

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1847,16 +1847,14 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
 }
 
-static bool VerifyTaprootCommitment(const std::vector<unsigned char>& control, const std::vector<unsigned char>& program, const CScript& script, uint256& tapleaf_hash)
+uint256 ComputeTapleafHash(uint8_t leaf_version, const CScript& script)
```
I mildly object the the use of the type `CScript` here.  While it is true that BIP-341 names this stack item "script", this value is only interpreted as a CScript in the case of BIP-342 where the `leaf_version` is equal to `TAPROOT_LEAF_TAPSCRIPT` (modulo `TAPROOT_LEAF_MASK`).  Otherwise it is simply a bytestring.
